### ISSUE TITLE
Serial killing performance on Pi Zero

### ISSUE THREAD

**Issue by pdeguing** [pyserial/pyserial#413]
Hello,

I'm trying to read sensor datas from an Arduino to a Pi Zero for a self-driving car project. I've stuck for more than a week now due to serial communication. I cannot get the Pi to read from usb via pyserial at reasonable speed.

My first guess was that the usb communication by itself was the bottleneck. But after running many tests, my conclusion is that the Pi is dying as soon as I open the serial port.

As soon as the script runs: `ser = serial.Serial('/dev/ttyUSB0', 74880, timeout=None)`, the Pi runs 50 times slower.

My last test is the most basic code possible. I have two scripts that runs a loop printing 1000 times a character. One contains `ser = serial.Serial('/dev/ttyUSB0', 74880, timeout=None)`, the second has the line commented out and an extra `sleep(1)` to approximate the time it takes to run the line (from approximative manual measurements).

On the arduino side, there is almost nothing. I setup the serial and prints "hello world" in the loop. It doesn't seem a problem specific to the arduino side since it runs perfectly when connected to the Mac or when running a serial program in C from the Pi.

Here are my two test scripts:

Test 1 with serial:

```python
import serial
from time import sleep

ser = serial.Serial('/dev/ttyUSB0', 74880, timeout=None)
sleep(2)
for i in range(0, 1000):
  print("x", end="")
```
Test 2 without serial:

```python
import serial
from time import sleep

#ser = serial.Serial('/dev/ttyUSB0', 74880, timeout=None)
sleep(3)
for i in range(0, 1000):
  print("x", end="")
```
When running time python3 for each, I get:

test 1: _from 10 to 30sec_
test 2: _around 4sec_

I've been looking into it for more than a week now and still did not find any workable solution or fix. I don't know what is wrong.

Thanks,

**Comment by harryberlin** [comment]
there is no ser.read or write in the code.
print is only for console output.
can handle the arduino your custom baudrate?

**Comment by pdeguing** [comment]
> there is no ser.read or write in the code.
> print is only for console output.
> can handle the arduino your custom baudrate?

Thank you for your answer. Indeed I removed everything I could from the code in order to identify the pain point. The problem is totally linked to me opening the serial device, so this one line. That's why there is nothing else in the code.

**Comment by harryberlin** [comment]
oh, you mean the init takes a long time.

**Comment by pdeguing** [comment]
> oh, you mean the init takes a long time.

No, I mean that as soon as the serial device is open, the Pi runs something like 50 times slower. It's running, but extremely slow. I tried another Pi and I have the same problem.

**Comment by harryberlin** [comment]
the same with standard baudrate?

**Comment by zsquareplusc** [comment]
can you check with "top" or some other CPU monitoring tool which process is having a high CPU load? is it the python script or maybe there is a modem manager or similar tool?

pySerial does not start a thread on it's own, so if just opening the port already causes CPU load, then I'd suspect either an other application also accessing the serial port at the same time or the driver has an issue :/

The raspi also only has one USB host so all the devices share the same bus. A serial dongle may be inefficient as it may use a lot of small packets. That may depend on the type of serial traffic and on the driver/firmware of the dongle. Though i would not expect that the CPU "feels" slower in that case but it may reduce network throughput (though the Pi Zero does not have network...).

**Comment by acciderbolina** [comment]
Hello, same problem here, having a simple script that read the Serial I have usage up to 20-30%, while similar program like ser2net are basically noise.

I will verify if this affect only the hardware serial (ttyAMA0) or also virtual serial (not sure if i can test both ttyUSB and ttyACM, I need to see what i have available)
